lib/internal/modules/esm/importmap.js

```diff
@@ -0,0 +1,135 @@
+'use strict';
+const path = require('path');
+const { readFileSync } = require('fs');
+const { pathToFileURL, isURL } = require('internal/url');
+const { setOwnProperty } = require('internal/util');
+const { JSONParse, ObjectEntries, SafeMap } = primordials;
+// @TODO add error codes
+// const { } = require('internal/errors').codes;
+
+class ImportMap {
+  #baseURL;
+  imports = new SafeMap();
+  scopes = new SafeMap();
+
+  constructor(raw, baseURL) {
+    this.#baseURL = baseURL;
+    processImportMap(this, baseURL, raw);
+  }
+
+  get baseURL() {
+    return this.#baseURL;
+  }
+
+  static load(filepath) {
+    let str;
+    try {
+      str = readFileSync(filepath);
+    } catch (err) {
+      if (err.code !== 'ENOENT') {
+        setOwnProperty(err, 'importMapPath', `${filepath}`);
+        throw err;
+      }
+      return;
+    }
+
+    let json;
+    try {
+      json = JSONParse(str);
+    } catch (err) {
+      setOwnProperty(err, 'importMap', `${str}`);
+      throw err;
+    }
+
+    return new ImportMap(json, pathToFileURL(path.dirname(filepath)));
+  }
+
+  resolve(specifier, parentURL) {
+    const scope = parentURL || this.baseURL;
+
+    // Process scopes
+    // @TODO this almost certainly needs to be a radix tree
+    for (const [prefix, mapping] of this.scopes) {
+      let mappedSpecifier = mapping.get(specifier);
+      if (scope.pathname.startsWith(prefix.pathname) && mappedSpecifier) {
+        if (!isURL(mappedSpecifier)) {
+          mappedSpecifier = pathToFileURL(mappedSpecifier, this.baseURL.pathname);
+          mapping.set(specifier, mappedSpecifier);
+        }
```

Comment:
I don't think we want to support paths in the import maps, only URLs

Reply:
The more I think about this I specifically think maybe I have reasoned out what you are meaning. Do you mean we shouldn't support remapping *to* bare specifiers?  So an example would be remapping `foo` to `foo-with-fix` and then letting the default resolution to handle it after that. I think it would be an unfortunate loss of functionality if we limited this, but I think it is likely something we should call out as an open question once I open the real PR.

Reply:
I think if the import maps spec doesn’t support it, then we shouldn’t either. You can always map `foo` to `./node_modules/foo-with-fix/whatever`; I feel like that’s probably what the spec would want us to do.

Reply:
> this is an alternative to the ESM resolution algorithm

I am not sure I agree with this assessment, but I am also not sure that it matters much for the technical details. IMO there is a lot of ergonomic things enabled by being able to map to a different bare specifier. The problem is that to do something simple like providing a patched package (my example) if we don't support this would mean an implementation which generates that import map would need to fully respect all of the node features like `exports` and `main` etc. Ideally I would like to enable this style of remapping without tooling authors having to relicate all that in their tools (or even import it from node).

Reply:
I am going to call this one resolved. We can have any further conversation in the main PR.